### ISSUE TITLE
Bug Report:  java.lang.LinkageError while classloading “com.nimbusds.jwt.JWTClaimsSet” FISH-5941

### ISSUE THREAD

**Issue by UlrichCech** [payara/Payara#5533]
# Description #
----------

I had a normal JavaEE application, which runs just fine with Java 11 and Payara 5.2021.4.

Now I wanted to update to Payara 5.2021.9 and also update Java to version 17 (LTS). But now, the application just starts fine, but if I try to make a first REST call, which involves authentication (where JWT token is involved), I get the attached error.

It seems to have something to do with the javax.ws.rs.container.ContainerRequestFilter, which is perhaps loaded in another classloader, which causes the error.

ADDITIONAL HINT: The problem was introduced since in Payara 5.2021.6 (all versions lower than this version work well, from version 5.2021.6 up to 5.2021.9), the error occurs.

## Expected Outcome ##

There is no java.lang.LinkageError or classloading problem.

## Current Outcome ##

```
javax.ejb.EJBException: loader constraint violation: loader org.apache.felix.framework.BundleWiringImpl$BundleClassLoader @6659885b wants to load class com.nimbusds.jwt.JWTClaimsSet. A different class with the same name was previously loaded by org.apache.felix.framework.BundleWiringImpl$BundleClassLoader @190fc9de. (com.nimbusds.jwt.JWTClaimsSet is in unnamed module of loader org.apache.felix.framework.BundleWiringImpl$BundleClassLoader @190fc9de, parent loader java.net.URLClassLoader @59906517)
	at com.sun.ejb.containers.EJBContainerTransactionManager.processSystemException(EJBContainerTransactionManager.java:723)
	at com.sun.ejb.containers.EJBContainerTransactionManager.completeNewTx(EJBContainerTransactionManager.java:652)
	at com.sun.ejb.containers.EJBContainerTransactionManager.postInvokeTx(EJBContainerTransactionManager.java:482)
	at com.sun.ejb.containers.BaseContainer.postInvokeTx(BaseContainer.java:4592)
	at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:2125)
	at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:2095)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:220)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandlerDelegate.invoke(EJBLocalObjectInvocationHandlerDelegate.java:90)
	at com.sun.proxy.$Proxy245.validateToken(Unknown Source)
	at de.test.platform.security.boundary.__EJB31_Generated__JWTTokenService__Intf____Bean__.validateToken(Unknown Source)
	at de.test.platform.security.boundary.AuthenticationFilter.validateToken(AuthenticationFilter.java:141)
	at de.test.platform.security.boundary.AuthenticationFilter.filter(AuthenticationFilter.java:97)
	at org.glassfish.jersey.server.ContainerFilteringStage.apply(ContainerFilteringStage.java:108)
	at org.glassfish.jersey.server.ContainerFilteringStage.apply(ContainerFilteringStage.java:44)
	at org.glassfish.jersey.process.internal.Stages.process(Stages.java:173)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:245)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:232)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:680)
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:394)
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:346)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:366)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:319)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:205)
	at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1636)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:259)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:161)
	at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:757)
	at org.apache.catalina.core.StandardPipeline.doChainInvoke(StandardPipeline.java:581)
	at com.sun.enterprise.web.WebPipeline.invoke(WebPipeline.java:97)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:158)
	at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:371)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:238)
	at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:520)
	at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:217)
	at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:182)
	at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:156)
	at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:218)
	at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:95)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:260)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:177)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:109)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:88)
	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:53)
	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:524)
	at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:89)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:94)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:33)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:114)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:569)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:549)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.LinkageError: loader constraint violation: loader org.apache.felix.framework.BundleWiringImpl$BundleClassLoader @6659885b wants to load class com.nimbusds.jwt.JWTClaimsSet. A different class with the same name was previously loaded by org.apache.felix.framework.BundleWiringImpl$BundleClassLoader @190fc9de. (com.nimbusds.jwt.JWTClaimsSet is in unnamed module of loader org.apache.felix.framework.BundleWiringImpl$BundleClassLoader @190fc9de, parent loader java.net.URLClassLoader @59906517)
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017)
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.defineClass(BundleWiringImpl.java:2344)
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.defineClassParallel(BundleWiringImpl.java:2162)
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.findClass(BundleWiringImpl.java:2096)
	at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1565)
	at org.apache.felix.framework.BundleWiringImpl.access$300(BundleWiringImpl.java:79)
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:1982)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at com.nimbusds.jwt.SignedJWT.getJWTClaimsSet(SignedJWT.java:101)
	at de.test.platform.security.boundary.JWTTokenService.validateToken(JWTTokenService.java:72)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.glassfish.ejb.security.application.EJBSecurityManager.runMethod(EJBSecurityManager.java:588)
	at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:408)
	at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4826)
	at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:665)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:834)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:615)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.doCall(SystemInterceptorProxy.java:163)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.aroundInvoke(SystemInterceptorProxy.java:140)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:888)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:833)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:615)
	at org.jboss.weld.module.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:72)
	at org.jboss.weld.module.ejb.SessionBeanInterceptor.aroundInvoke(SessionBeanInterceptor.java:52)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:888)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:833)
	at com.sun.ejb.containers.interceptors.InterceptorManager.intercept(InterceptorManager.java:375)
	at com.sun.ejb.containers.BaseContainer.__intercept(BaseContainer.java:4798)
	at com.sun.ejb.containers.BaseContainer.intercept(BaseContainer.java:4786)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:212)
	... 50 more
]]
```

## Steps to reproduce ## 

1. checkout the linked "reproducer project"
 https://github.com/UlrichCech/payara-jwt-reproducer/tree/Payara-5.2021.9_NimbusJoseJwt_ClassloadingProblem
   on branch "Payara-5.2021.9_NimbusJoseJwt_ClassloadingProblem"
2. run the testcase JWTTokenServiceTest#generateJWT()  to produce a valid JWT token
3. curl -v -H "Authorization: Bearer *JWT-Token* " http://localhost:8080/api/v1/users/123

## Environment ##

- **Distribution: Server Full Profile - Version 5.2021.4 - 5.2021.9
- **JDK Version: OpenJDK 11 and 17
- **Operating System: Mac

**Comment by OndroMih** [comment]
I easily reproduced the issue. It looks like it happens because nimbus classes are contained in 2 separate Payara modules in glassfish/modules directory: nimbus-jose-jwt.jar and also in security-connector-oidc-client.jar. This needs to be fixed so that there are no duplicate JWT classes.

**Comment by luiseufrasio** [comment]
I also reproduced it locally. 
I ran also with payara-micro-5.2021.10.jar and in this case the error does not occur.

**Comment by JEFROG** [comment]
I have the same error:

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">

<head>
	<title>Payara Server 5.2022.1 #badassfish - Error report</title>
	<style type="text/css">
		
	</style>
</head>

<body>
	<h1>HTTP Status 500 - Internal Server Error</h1>
	<hr />
	<p><b>type</b> Exception report</p>
	<p><b>message</b>Internal Server Error</p>
	<p><b>description</b>The server encountered an internal error that prevented it from fulfilling this request.</p>
	<p><b>exception</b>
		<pre>javax.servlet.ServletException: org.glassfish.jersey.server.ContainerException: java.lang.LinkageError: loader constraint violation: loader org.glassfish.web.loader.WebappClassLoader @41d73560 wants to load class com.nimbusds.jwt.JWTClaimsSet. A different class with the same name was previously loaded by org.apache.felix.framework.BundleWiringImpl$BundleClassLoader @66a857cd. &#40;com.nimbusds.jwt.JWTClaimsSet is in unnamed module of loader org.apache.felix.framework.BundleWiringImpl$BundleClassLoader @66a857cd, parent loader java.net.URLClassLoader @73a8dfcc&#41;</pre>
	</p>
	<p><b>root cause</b>
		<pre>org.glassfish.jersey.server.ContainerException: java.lang.LinkageError: loader constraint violation: loader org.glassfish.web.loader.WebappClassLoader @41d73560 wants to load class com.nimbusds.jwt.JWTClaimsSet. A different class with the same name was previously loaded by org.apache.felix.framework.BundleWiringImpl$BundleClassLoader @66a857cd. &#40;com.nimbusds.jwt.JWTClaimsSet is in unnamed module of loader org.apache.felix.framework.BundleWiringImpl$BundleClassLoader @66a857cd, parent loader java.net.URLClassLoader @73a8dfcc&#41;</pre>
	</p>
	<p><b>root cause</b>
		<pre>java.lang.LinkageError: loader constraint violation: loader org.glassfish.web.loader.WebappClassLoader @41d73560 wants to load class com.nimbusds.jwt.JWTClaimsSet. A different class with the same name was previously loaded by org.apache.felix.framework.BundleWiringImpl$BundleClassLoader @66a857cd. &#40;com.nimbusds.jwt.JWTClaimsSet is in unnamed module of loader org.apache.felix.framework.BundleWiringImpl$BundleClassLoader @66a857cd, parent loader java.net.URLClassLoader @73a8dfcc&#41;</pre>
	</p>
	<p><b>note</b> <u>The full stack traces of the exception and its root causes are available in the Payara Server
			5.2022.1 #badassfish logs.</u></p>
	<hr />
	<h3>Payara Server 5.2022.1 #badassfish</h3>
</body>

</html>

**Comment by shub8968** [comment]
HI @JEFROG,

It will be fixed in the upcoming release.

Thanks,
Shubham

**Comment by OndroMih** [comment]
This is fixed now in 5.2022.2: https://docs.payara.fish/community/docs/release-notes/release-notes-2022-2.html
